File: src/main/java/com/dasoops/common/cache/BaseCache.java
package com.dasoops.common.cache;

import cn.hutool.core.lang.func.VoidFunc0;
import com.dasoops.common.entity.enums.base.IRedisHashKeyEnum;
import com.dasoops.common.entity.enums.base.IRedisKeyEnum;
import com.dasoops.common.entity.enums.base.PrefixRedisKeyShamEnum;
import org.springframework.dao.DataAccessException;
import org.springframework.data.redis.connection.DataType;
import org.springframework.data.redis.connection.RedisConnection;
import org.springframework.data.redis.core.*;
import com.alibaba.fastjson2.JSON;

import java.util.HashMap;
import java.util.List;
import java.util.Map;
import java.util.Set;
import java.util.concurrent.TimeUnit;
import java.util.function.Function;
import java.util.stream.Collectors;

/**
 * @title BaseCache
 * @classPath com.dasoops.common.cache.BaseCache
 * @author DasoopsNicole@Gmail.com
 * @date 2023/01/01
 * @version 1.0.0
 * @description 基地缓存
 */
public class BaseCache {

    private final StringRedisTemplate redis;

    public BaseCache(StringRedisTemplate redis) {
        this.redis = redis;
    }

    /* -- Operations Begin -- */

    private ValueOperations<String, String> value() {
        return redis.opsForValue();
    }

    private ListOperations<String, String> list() {
        return redis.opsForList();
    }

    private SetOperations<String, String> set() {
        return redis.opsForSet();
    }

    private ZSetOperations<String, String> zSet() {
        return redis.opsForZSet();
    }

    private HashOperations<String, String, String> hash() {
        return redis.opsForHash();
    }

    /* -- Operations End -- */

    /* -- Global Begin -- */

    protected Set<String> keys() {
        return redis.keys("*");
    }

    protected Set<String> keys(String keyword) {
        return redis.keys("*" + keyword + "*");
    }

    protected Set<String> keys4Prefix(IRedisKeyEnum redisKeyEnum) {
        return redis.keys(redisKeyEnum.getKey() + "*");
    }

    protected void remove(IRedisKeyEnum redisKeyEnum) {
        redis.delete(redisKeyEnum.getKey());
    }

    protected void remove4Prefix(IRedisKeyEnum redisKeyEnum) {
        Set<String> keys = this.keys(redisKeyEnum.getKey() + "*");
        if (keys != null && keys.size() > 0) {
            redis.delete(keys);
        }
    }

    protected void expire(IRedisKeyEnum redisKeyEnum, Long timeout, TimeUnit timeUnit) {
        redis.expire(redisKeyEnum.getKey(), timeout, timeUnit);
    }

    protected void expire4Prefix(IRedisKeyEnum redisKeyEnum, Long timeout, TimeUnit timeUnit) {
        Set<String> keys = redis.keys(redisKeyEnum.getKey() + "*");
        if (keys == null) {
            return;
        }
        keys.forEach(key -> redis.expire(key, timeout, timeUnit));
    }

    protected DataType type(IRedisKeyEnum redisKeyEnum) {
        return redis.type(redisKeyEnum.getKey());
    }

    protected String getJSONString(IRedisKeyEnum redisKeyEnum) {
        DataType type = type(redisKeyEnum);
        switch (type) {
            case SET -> {
                return JSON.toJSONString(members(redisKeyEnum));
            }
            case HASH -> {
                return JSON.toJSONString(entries(redisKeyEnum));
            }
            case LIST -> {
                return JSON.toJSONString(lget(redisKeyEnum));
            }
            case STRING -> {
                return get(redisKeyEnum);
            }
            default -> {
                return null;
            }
        }
    }

    protected void transaction(VoidFunc0 func) {
        redis.setEnableTransactionSupport(true);
        redis.multi();

        func.callWithRuntimeException();

        redis.exec();
        redis.setEnableTransactionSupport(false);
    }

    protected Boolean hasKey(IRedisKeyEnum redisKey) {
        return redis.hasKey(redisKey.getKey());
    }
    /* -- Global End -- */

    /* -- Value Begin -- */

    protected void set(IRedisKeyEnum redisKeyEnum, String value) {
        value().set(redisKeyEnum.getKey(), value);
    }

    protected String get(IRedisKeyEnum redisKeyEnum) {
        return value().get(redisKeyEnum.getKey());
    }

    protected String getAndDelete(IRedisKeyEnum redisKeyEnum) {
        return value().getAndDelete(redisKeyEnum.getKey());
    }

    protected void setAndExpire(IRedisKeyEnum redisKeyEnum, String value, Long time, TimeUnit timeUnit) {
        value().set(redisKeyEnum.getKey(), value);
        expire(redisKeyEnum, time, timeUnit);
    }
    /* -- Value End -- */

    /* -- Hash Begin -- */

    protected Map<String, String> entries(IRedisKeyEnum redisKeyEnum) {
        return hash().entries(redisKeyEnum.getKey());
    }

    protected Map<String, Map<String, String>> entries4Prefix(PrefixRedisKeyShamEnum redisKeyShamEnum) {
        Set<String> keySet = this.keys(redisKeyShamEnum.getKey());
        return keySet.stream().collect(Collectors.toMap(key -> key, key -> hash().entries(key)));
    }

    protected String hget(IRedisKeyEnum redisKeyEnum, IRedisHashKeyEnum redisHashKeyEnum) {
        return hash().get(redisKeyEnum.getKey(), redisHashKeyEnum.getKey());
    }

    protected String hget(IRedisKeyEnum redisKeyEnum, String redisHashKey) {
        return hash().get(redisKeyEnum.getKey(), redisHashKey);
    }

    protected void hset(IRedisKeyEnum redisKeyEnum, IRedisHashKeyEnum redisHashKeyEnum, String value) {
        hash().put(redisKeyEnum.getKey(), redisHashKeyEnum.getKey(), value);
    }

    protected void hset(IRedisKeyEnum redisKeyEnum, String hashKey, String value) {
        hash().put(redisKeyEnum.getKey(), hashKey, value);
    }

    protected void hset(IRedisKeyEnum redisKeyEnum, Map<String, String> valueMap) {
        hash().putAll(redisKeyEnum.getKey(), valueMap);
    }

    protected boolean hhasKey(IRedisKeyEnum redisKeyEnum, String hashKey) {
        return hash().hasKey(redisKeyEnum.getKey(), hashKey);
    }

    protected String hgetAndDelete(IRedisKeyEnum redisKeyEnum, String hashKey) {
        boolean b = this.hhasKey(redisKeyEnum, hashKey);
        if (!b) {
            return null;
        }
        String str = hash().get(redisKeyEnum.getKey(), hashKey);
        hdelete(redisKeyEnum, hashKey);
        return str;
    }

    protected void hdelete(IRedisKeyEnum redisKeyEnum, String hashKey) {
        hash().delete(redisKeyEnum.getKey(), hashKey);
    }

    /* -- Hash End -- */

    /* -- List Begin -- */

    protected void lset(IRedisKeyEnum redisKeyEnum, List<String> valueList) {
        list().rightPushAll(redisKeyEnum.getKey(), valueList);
    }

    protected void lset(IRedisKeyEnum redisKeyEnum, String value) {
        list().rightPush(redisKeyEnum.getKey(), value);
    }

    protected List<String> lget(IRedisKeyEnum redisKeyEnum) {
        return list().range(redisKeyEnum.getKey(), 0, -1);
    }

    protected String lget(IRedisKeyEnum redisKeyEnum, Integer index) {
        return list().index(redisKeyEnum.getKey(), index);
    }

    /* -- List End -- */

    /* -- Set Begin -- */

    protected void sadd(IRedisKeyEnum redisKeyEnum, List<String> valueList) {
        set().add(redisKeyEnum.getKey(), valueList.toArray(new String[0]));
    }

    protected void sadd(IRedisKeyEnum redisKeyEnum, Set<String> valueSet) {
        set().add(redisKeyEnum.getKey(), valueSet.toArray(new String[0]));
    }

    protected void saddAndExpire(IRedisKeyEnum redisKeyEnum, Set<String> valueSet, Long time, TimeUnit timeUnit) {
        this.transaction(() -> {
            set().add(redisKeyEnum.getKey(), valueSet.toArray(new String[0]));
            redis.expire(redisKeyEnum.getKey(), time, timeUnit);
        });
    }

    protected void sadd(IRedisKeyEnum redisKeyEnum, String value) {
        set().add(redisKeyEnum.getKey(), value);
    }

    protected Set<String> members(IRedisKeyEnum redisKeyEnum) {
        return set().members(redisKeyEnum.getKey());
    }

    /**
     * 根据前缀获取所有项并分组
     *
     * @param redisKeyEnum 复述,关键枚举
     * @param groupingKeyConvertFunction 分组键转换函数
     * @param valueConvertFunction 分组值转换函数
     * @return {@link Map}<{@link R1}, {@link Set}<{@link R2}>>
     */
    protected <R1, R2> Map<R1, Set<R2>> sGetGroupingByPrefix(IRedisKeyEnum redisKeyEnum, Function<String, R1> groupingKeyConvertFunction, Function<String, R2> valueConvertFunction) {
        Set<String> keyStringSet = this.keys4Prefix(redisKeyEnum);
        HashMap<R1, Set<R2>> resultMap = new HashMap<>(keyStringSet.size());
        for (String key : keyStringSet) {
            Set<String> members = set().members(key);
            if (members == null) {
                return resultMap;
            }
            Set<R2> set = members.stream().map(valueConvertFunction).collect(Collectors.toSet());
            String keyString = key.substring(key.lastIndexOf(":") + 1);
            R1 groupingKey = groupingKeyConvertFunction.apply(keyString);
            resultMap.put(groupingKey, set);
        }
        return resultMap;
    }

    protected Map<String, Set<String>> sGetGroupingByPrefix(IRedisKeyEnum redisKeyEnum) {
        return this.sGetGroupingByPrefix(redisKeyEnum, string -> string, string -> string);
    }

    /* -- Set End -- */
}

File: src/main/java/com/dasoops/common/cache/BaseCache.kt
package com.dasoops.common.cache

import cn.hutool.core.util.RandomUtil
import cn.hutool.core.util.StrUtil
import cn.hutool.core.util.TypeUtil
import com.dasoops.common.entity.enums.cache.ICacheKeyEnum
import com.dasoops.common.exception.CacheException
import com.dasoops.common.util.json.toJsonStr
import com.dasoops.common.task.AutoInit
import kotlinx.coroutines.runBlocking
import kotlinx.coroutines.withTimeoutOrNull
import org.springframework.data.redis.connection.DataType
import org.springframework.data.redis.core.StringRedisTemplate
import java.util.concurrent.TimeUnit
import java.util.function.Function
import java.util.stream.Collectors

/**
 * @author DasoopsNicole@Gmail.com
 * @version 1.0.0
 * @title BaseCache
 * @classPath com.dasoops.common.cache.BaseCache
 * @date 2023/01/01
 * @description 基地缓存
 */
abstract class BaseCache<E : ICacheKeyEnum> : RedisOperations(), ICache, AutoInit {

    /**
     * 设置操作类
     * @param [template] 操作类
     */
    fun setTemplate(template: StringRedisTemplate) {
        super.redis = template
    }

    /* -- Global Begin -- */
    /**
     * 缓存初始化
     * 调用init方法
     */
    override fun init() {
        clear()
    }

    /**
     * 清除缓存数据
     */
    protected fun clear() {
        //获取泛型
        @Suppress("UNCHECKED_CAST")
        val typeArgument = TypeUtil.getTypeArgument(this::class.java) as Class<E>
        //获取泛型实例数组
        val enumConstantArray = typeArgument.enumConstants
        //清除
        enumConstantArray.forEach(this::remove4Prefix)
    }

    protected fun keys(): Collection<String>? {
        return super.keys("*")
    }

    protected fun keys(cacheKeyEnum: E): Collection<String>? {
        return super.keys(cacheKeyEnum.getKey())
    }

    protected fun keys4Prefix(key: String): Collection<String>? {
        return super.keys(StrUtil.addSuffixIfNot(key, "*"))
    }

    protected fun keys4Prefix(cacheKeyEnum: E): Collection<String>? {
        return this.keys4Prefix(cacheKeyEnum.getKey())
    }

    protected fun remove(cacheKeyEnum: E) {
        super.remove(cacheKeyEnum.getKey())
    }

    protected fun remove4Prefix(key: String) {
        super.keys(key)?.let {
            super.remove(*it.toTypedArray())
        }
    }

    protected fun remove4Prefix(cacheKeyEnum: E) {
        this.remove4Prefix(cacheKeyEnum.getKey())
    }

    protected fun expire(cacheKeyEnum: E, timeout: Long, timeUnit: TimeUnit) {
        super.expire(cacheKeyEnum.getKey(), timeout, timeUnit)
    }

    protected fun expire4Prefix(cacheKeyEnum: E, timeout: Long, timeUnit: TimeUnit) {
        super.expire4Prefix(cacheKeyEnum.getKey(), timeout, timeUnit)
    }

    protected fun type(cacheKeyEnum: E): DataType {
        return super.type(cacheKeyEnum.getKey())
    }

    protected fun getJsonString(key: String): String? {
        return when (type(key)) {
            DataType.SET -> members(key)?.toJsonStr()
            DataType.HASH -> entries(key)?.toJsonStr()
            DataType.LIST -> list(key)?.toJsonStr()
            DataType.STRING -> get(key)?.toJsonStr()
            else -> throw CacheException.UNDEFINED_CAST.get()
        }
    }

    protected fun getJsonString(cacheKeyEnum: E): String? {
        return this.getJsonString(cacheKeyEnum.getKey())
    }

    protected fun hasKey(cacheKeyEnum: E): Boolean {
        return super.hasKey(cacheKeyEnum.getKey())
    }

    /* -- Value Begin -- */
    protected fun set(cacheKeyEnum: E, value: String) {
        super.set(cacheKeyEnum.getKey(), value)
    }

    protected fun get(cacheKeyEnum: E): String? {
        return super.get(cacheKeyEnum.getKey())
    }

    protected fun getAndDelete(cacheKeyEnum: E): String? {
        return super.getAndDelete(cacheKeyEnum.getKey())
    }

    protected fun setAndExpire(cacheKeyEnum: E, value: String, time: Long, timeUnit: TimeUnit) {
        super.setAndExpire(cacheKeyEnum.getKey(), value, time, timeUnit)
    }

    protected fun lock(cacheKeyEnum: E, value: String, time: Long, timeUnit: TimeUnit) {
        this.lock(cacheKeyEnum.getKey(), value, time, timeUnit)
    }

    protected fun lock(key: String, value: String, time: Long, timeUnit: TimeUnit) = runBlocking {
        // 上锁 成功直接返回
        // 失败50ms - 100ms 重试, 随机数意图是错开尝试时间
        // 超时时间2000ms 超时直接抛异常
        withTimeoutOrNull(2000) {
            repeat(RandomUtil.randomInt(50, 100)) {
                if (super.setIfAbsent(key, value, time, timeUnit)) {
                    return@withTimeoutOrNull
                }
            }
        } ?: throw CacheException.GET_LOCK_ERROR.get()

    }

    protected fun unlock(key: String) {
        this.remove(key)
    }

    protected fun unlock(cacheKeyEnum: E) {
        this.remove(cacheKeyEnum.getKey())
    }

    /* -- Hash Begin -- */
    protected fun entries(cacheKeyEnum: E): Map<String, String>? {
        return super.entries(cacheKeyEnum.getKey())
    }

    protected fun entries4Prefix(cacheKeyEnum: E): Map<String, Map<String, String>>? {
        return super.entries4Prefix(cacheKeyEnum.getKey())
    }

    protected fun hget(cacheKeyEnum: E, hashKey: String): String? {
        return super.hget(cacheKeyEnum.getKey(), hashKey)
    }

    protected fun hput(cacheKeyEnum: E, hashKey: String, value: String) {
        super.hput(cacheKeyEnum.getKey(), hashKey, value)
    }

    protected fun hputAll(cacheKeyEnum: E, valueMap: Map<String, String>) {
        super.hputAll(cacheKeyEnum.getKey(), valueMap)
    }

    protected fun hhasKey(cacheKeyEnum: E, hashKey: String): Boolean {
        return super.hhasKey(cacheKeyEnum.getKey(), hashKey)
    }

    protected fun hgetAndDelete(cacheKeyEnum: E, hashKey: String): String? {
        return this.hgetAndDelete(cacheKeyEnum.getKey(), hashKey)
    }

    protected fun hgetAndDelete(key: String, hashKey: String): String? {
        val str = super.hget(key, hashKey) ?: return null
        hdelete(key, hashKey)
        return str
    }

    protected fun hdelete(cacheKeyEnum: E, hashKey: String) {
        super.hdelete(cacheKeyEnum.getKey(), hashKey)
    }

    /* -- List Begin -- */
    protected fun lset(cacheKeyEnum: E, valueList: Collection<String>) {
        super.lrightPushAll(cacheKeyEnum.getKey(), valueList)
    }

    protected fun ladd(cacheKeyEnum: E, value: String) {
        super.lrightPush(cacheKeyEnum.getKey(), value)
    }

    protected fun list(cacheKeyEnum: E): Collection<String>? {
        return super.list(cacheKeyEnum.getKey())
    }

    protected fun lget(cacheKeyEnum: E, index: Int): String? {
        return super.lget(cacheKeyEnum.getKey(), index)
    }

    protected fun lget(cacheKeyEnum: E): String? {
        return super.lget(cacheKeyEnum.getKey(), 0)
    }

    /* -- Set Begin -- */

    protected fun sadd(key: String, valueSet: Collection<String>) {
        super.sadd(key, *valueSet.toTypedArray())
    }

    protected fun sadd(cacheKeyEnum: E, valueSet: Collection<String>) {
        this.sadd(cacheKeyEnum.getKey(), valueSet)
    }

    protected fun sremove(key: String, value: String) {
        super.sremove(key, value)
    }

    protected fun sremove(cacheKeyEnum: E, value: String) {
        this.sremove(cacheKeyEnum.getKey(), value)
    }

    protected fun sremove(key: String, value: Collection<String>) {
        super.sremove(key, *value.toTypedArray())
    }

    protected fun sremove(cacheKeyEnum: E, value: Collection<String>) {
        super.sremove(cacheKeyEnum.getKey(), *value.toTypedArray())
    }

    protected fun saddAndExpire(key: String, valueSet: Collection<String>, time: Long, timeUnit: TimeUnit) {
        transaction {
            super.sadd(key, *valueSet.toTypedArray())
            super.expire(key, time, timeUnit)
        }
    }

    protected fun saddAndExpire(cacheKeyEnum: E, valueSet: Collection<String>, time: Long, timeUnit: TimeUnit) {
        this.saddAndExpire(cacheKeyEnum.getKey(), valueSet, time, timeUnit)
    }

    protected fun members(cacheKeyEnum: E): Collection<String>? {
        return super.members(cacheKeyEnum.getKey())
    }

    protected fun <R1, R2> members4Prefix(
        key: String,
        keyConvertFunction: Function<String, R1>,
        valueConvertFunction: Function<String, R2>
    ): Map<R1, Collection<R2>>? {
        val keySet = keys4Prefix(key) ?: return null
        return keySet.stream().map {
            val members = super.members(it) ?: return@map null
            val resValue = members.stream().map(valueConvertFunction).collect(Collectors.toSet())
            val resKey = keyConvertFunction.apply(it.substringAfterLast(":"))
            resKey to resValue
        }.filter { it != null }.collect(Collectors.toMap({ it!!.first }, { it!!.second }))
    }

    protected fun <R1, R2> members4Prefix(
        cacheKeyEnum: E,
        keyConvertFunction: Function<String, R1>,
        valueConvertFunction: Function<String, R2>
    ): Map<R1, Collection<R2>>? {
        return this.members4Prefix(cacheKeyEnum.getKey(), keyConvertFunction, valueConvertFunction)
    }

    protected fun members4Prefix(cacheKeyEnum: E): Map<String, Collection<String>>? {
        return this.members4Prefix(cacheKeyEnum.getKey())
    }

    protected fun members4Prefix(key: String): Map<String, Collection<String>>? {
        return this.members4Prefix(key, { string: String -> string }) { string: String -> string }
    }
}